src/library/scala/collection/mutable/ArrayBuilder.scala

```diff
@@ -27,20 +28,31 @@ sealed abstract class ArrayBuilder[T]
   protected[this] def elems: Array[T]
   protected var size: Int = 0
 
+  def MaxSize: Int = PStatics.VM_MaxArraySize
```

Comment:
Is this portable and does it work on all Scala version?  For instance, if I try it in the REPL in 3.3.0 with JDK 17.0.7, I get a ReferenceError.

`jdk.internal.util.ArraysSupport.SOFT_MAX_ARRAY_LENGTH` is another source for this value.

But there's a pretty long tradition of just assuming that Int.MaxValue - 8 is safe, e.g.

https://hg.openjdk.org/jdk7/jdk7/jdk/file/tip/src/share/classes/java/util/ArrayList.java#l190

Anyway, this is hardly the only place where we have to worry about bounds overflows.

Can you document why this is the correct choice?

Reply:
`ArrayBuffer` uses this value and it is definitely wrong on some JVMs.  A better solution would be to change `PStatics.VM_MaxArraySize`'s value to `Int.MaxValue - 8`

Reply:
I can't access sPStatics.VM_MaxArraySizes from code outside of the scala/scala repository, so not surprised that it didn't work in a REPL.

`SOFT_MAX_ARRAY_LENGTH` sounds like it could replace PStatics.  I'll look into that.

Reply:
> `jdk.internal.util.ArraysSupport.SOFT_MAX_ARRAY_LENGTH` is another source for this value.

![image](https://github.com/scala/scala/assets/24378760/074024b5-41db-4ff2-9801-fd704c9a7cb3)

Maybe we should just change `PStatics.VM_MaxArraySize` to `Int.MaxInt - 8`?.

Reply:
@sjrd suggested hard coding a `private final val MaxArraySize:Int = Int.MaxInt - 8` so the newest version uses that.